[Split de #2325] Migración de consumidor HTTP a safeRequest + QA E2E end-to-end

## Contexto

Historia hija de #2325 (Split de #2318, raíz #2314). Entrega la **validación end-to-end** del módulo defensivo: migra al menos un consumidor HTTP de alto riesgo a `safeRequest`, refina los mensajes de error finales, activa el logging de denials y cubre el QA E2E con video.

## Alcance

Implementa CA9, CA10 (refinamiento final), CA11 (logging en producción) y CA14.

### Entregables

#### Migración de consumidor HTTP

- [ ] Migrar al menos UN consumidor HTTP del pipeline a `safeRequest`. Candidatos (elegir según criterio de riesgo):
  - `servicio-telegram.js` (recomendado — recibe URLs de usuarios vía Telegram, vector SSRF real)
  - `multimedia.js` (fetch a URLs externas dinámicas)
  - `listener-telegram.js`
  - `pulpo.js`
  - `dashboard-v2.js`
  - `test-condenser.js`
- [ ] El consumidor migrado NO debe perder funcionalidad existente: toda URL actualmente legítima debe estar en `endpoint-whitelist.json`.
- [ ] Inventario final del whitelist JSON contempla TODOS los endpoints usados por el consumidor migrado (verificación manual antes del PR).

#### Logging de denials

- [ ] Cuando `safeRequest` rechaza una URL, se loggea con: URL, host resuelto, IP resuelta, razón, stack.
- [ ] El logging pasa por el sanitizer de #2324 si está mergeado; si no, el módulo NO loggea headers sensibles por diseño (Authorization, Cookie, body).
- [ ] Formato consistente con el resto del pipeline (timestamp + nivel + componente).

#### DX final de mensajes (refinamiento)

- [ ] Revisar los mensajes del módulo contra el feedback real del consumidor migrado: los mensajes deben ser accionables en el contexto de producción, no solo en tests.
- [ ] Confirmar que ningún mensaje filtra secretos al integrar con el consumidor (validación manual con un header `Authorization` dummy).

#### QA E2E con video

- [ ] Video que demuestra:
  - El linter corriendo en pre-commit y bloqueando un commit con `rejectUnauthorized: false`
  - La anotación `::error` apareciendo inline en un PR de prueba de CI
  - Un intento de conexión a `http://169.254.169.254` desde el consumidor migrado siendo bloqueado con mensaje completo (URL, host, IP, razón, whitelist)
  - El guard de `NODE_TLS_REJECT_UNAUTHORIZED=0` abortando el proceso con mensaje accionable
  - El consumidor migrado funcionando normalmente con endpoints whitelisteados (Telegram envía mensajes, etc.)
- [ ] Label `qa:passed` antes de merge (nunca `qa:skipped`: historia tiene impacto de seguridad crítico).

### Fuera de alcance

- ❌ Módulo `endpoint-whitelist.js` core (Split A de #2325 — #2330)
- ❌ Linter + pre-commit + CI + docs (Split B de #2325 — #2331)
- ❌ Migrar los otros 5 consumidores HTTP restantes (follow-up en backlog; Split C solo requiere UNO).

## Dependencias

- **Hard:** Split A de #2325 (#2330) — necesita `safeRequest`/`assertAllowed` exportados.
- **Hard:** Split B de #2325 (#2331) — el linter debe estar activo para garantizar que la migración no introduce patterns prohibidos (ejemplo: un `fetch` directo bypaseando `safeRequest`).
- **Soft:** #2324 (sanitizer) — si está mergeado cuando arranca Split C, los logs de denials pasan por el sanitizer. Si no, logging básico sin headers sensibles.

## Testing

- QA E2E con video obligatorio (CA14).
- Smoke test del consumidor migrado: flujo normal funciona (ej: si se eligió `servicio-telegram.js`, mandar un mensaje Telegram de prueba y verificar que llega).
- Test adversarial: intento manual de SSRF al metadata AWS desde el consumidor → bloqueado con mensaje completo.

## Referencias

- Historia padre split: #2325
- Historia abuela: #2318
- Historia raíz: #2314
- Split A: #2330
- Split B: #2331

## ✅ Code Review — APROBADO

**Rama revisada:** `agent/2332-android-dev` (HEAD `c0c94bb4`, incluye merge de `agent/2332-backend-dev`)

### Archivos revisados (+280 / -66)

| Archivo | Cambio |
|---------|--------|
| `.pipeline/servicio-telegram.js` | Migrado de `https.request` directo a `httpClient.postJson` + `httpClient.request` (multipart). Agrega `logDenialIfAny` local. |
| `.pipeline/lib/http-client.js` | Nuevo try/catch alrededor del SSRF guard que emite `DENIAL <CODE> url=... host=... razon=... stack=...` antes de propagar. |
| `.pipeline/lib/redact.js` | Tercer paso en `redactUrlLike`: regex `/\/bot[^\/?#\s]{20,}/gi` → `/bot[REDACTED]` para cerrar fuga de `BOT_TOKEN` en path segments. |
| `.pipeline/lib/__tests__/redact.test.js` | +2 tests CA-11.1 (URL canónica + variantes / falsos positivos / case-insensitive). |
| `.pipeline/lib/__tests__/http-client.test.js` | +4 tests CA-11 (metadata AWS, loopback, happy path sin denial, DNS rebind de Telegram con `_dnsResolver` fake). |

### Checklist del reviewer

- ✅ **Código limpio y legible** — funciones pequeñas, comentarios apuntando a CAs y al issue
- ✅ **Patrones del pipeline respetados** — uso consistente de `httpClient.postJson` / `request`, `agentTag`, `ERROR_CODES` de `constants.js`
- ✅ **Sin código muerto ni TODOs** — `require('https')` eliminado, no quedan referencias al fetch inseguro
- ✅ **Logging estructurado** — nivel `ERROR`, tag del agente, campos url/host/razón/stack como exige CA-11
- ✅ **Sin secrets hardcodeados** — `BOT_TOKEN` sigue viniendo de config y ahora se redacta en todos los consumidores de `redactUrlLike` (DENIAL, info log, error, stack, sensitive)
- ✅ **Tests en español con códigos CA** — nomenclatura consistente con el resto de la suite (`CA-11 · ...`, `CA-11.1 · ...`)
- ✅ **Sin imports innecesarios** — `require('https')` removido de `servicio-telegram.js`
- ✅ **Scope acotado** — sólo el fix del rebote + la migración del CA-9; los otros 5 consumidores (pulpo, multimedia, listener-telegram, dashboard-v2, test-condenser) quedan explícitamente como follow-up según el body del issue

### Verificación funcional reproducida

```bash
$ node --test .pipeline/lib/__tests__/redact.test.js
  tests 15  pass 15  fail 0  skipped 0

$ node --test .pipeline/lib/__tests__/http-client.test.js
  tests 23  pass 23  fail 0  skipped 0
```

DENIAL real capturado en stderr durante el test de DNS rebind:
```
[ERROR] [http-client/test-ca11-telegram] DENIAL ERR_SSRF_BLOCKED
  url=https://api.telegram.org/bot[REDACTED]/sendMessage
  host=api.telegram.org
  razon=[HTTP_SSRF_BLOCKED]: IP privada 127.0.0.1 rechazada por política de seguridad (CA-9) ...
  stack=Error: ... | at buildSSRFError | at validateHostname | at doSingleRequest
```
Token literal `1234567890:ABCDefGHIjklMNOpqrsTUVwxyz_sensitive` **ausente** de stderr. ✓

### Observaciones (no-bloqueantes)

- El regex `/\/bot[^\/?#\s]{20,}/gi` es específico del formato Telegram (`/bot` + token opaco ≥ 20 chars). Si más adelante otros servicios usan prefijos similares (`/api`, `/key`, etc.), conviene generalizar con una allowlist de path-secret patterns. Queda como follow-up de hardening.
- Los GAPs del CA-14 (linter pre-commit, anotación `::error` en CI, guard `NODE_TLS_REJECT_UNAUTHORIZED=0`) dependen de #2331 (CLOSED NOT_PLANNED) y ya fueron aceptados por PO en la pasada previa.

**Recomendación:** mergear a `main` una vez que el gate de `qa:passed` esté asentado.

--- a/.pipeline/lib/__tests__/http-client.test.js
+++ b/.pipeline/lib/__tests__/http-client.test.js
@@ -442,3 +442,84 @@ test('postJson · serializa y setea content-type', withPermissiveSSRF(async () =
         await srv.close();
     }
 }));
+
+// ---- Tests CA-11 (issue #2332) · logging de denials SSRF --------------------
+
+/**
+ * Captura stderr durante la ejecución de `fn` y devuelve lo escrito.
+ * Usado para validar que el logger.error del http-client emitió el DENIAL.
+ */
+async function captureStderr(fn) {
+    const originalWrite = process.stderr.write.bind(process.stderr);
+    const chunks = [];
+    process.stderr.write = (chunk, enc, cb) => {
+        chunks.push(typeof chunk === 'string' ? chunk : chunk.toString('utf8'));
+        if (typeof enc === 'function') enc();
+        else if (typeof cb === 'function') cb();
+        return true;
+    };
+    try {
+        await fn();
+    } finally {
+        process.stderr.write = originalWrite;
+    }
+    return chunks.join('');
+}
+
+test('CA-11 · SSRF a 169.254.169.254 logea DENIAL con URL/host/razón/stack', async () => {
+    let err = null;
+    const stderr = await captureStderr(async () => {
+        try {
+            await get('http://169.254.169.254/latest/meta-data/', {
+                timeout: 5000,
+                agentTag: 'test-ca11',
+            });
+        } catch (e) {
+            err = e;
+        }
+    });
+
+    // El error sigue propagándose (no silenciamos denial).
+    assert.ok(err, 'el request debía fallar');
+    assert.equal(err.code, 'ERR_SSRF_BLOCKED');
+
+    // El log estructurado fue escrito a stderr con todos los campos requeridos.
+    assert.match(stderr, /ERROR/, 'debe ser nivel ERROR');
+    assert.match(stderr, /DENIAL ERR_SSRF_BLOCKED/);
+    assert.match(stderr, /url=http:\/\/169\.254\.169\.254/);
+    assert.match(stderr, /host=169\.254\.169\.254/);
+    assert.match(stderr, /razon=/);
+    assert.match(stderr, /stack=/);
+    assert.match(stderr, /test-ca11/); // agentTag en el tag del logger
+});
+
+test('CA-11 · SSRF a 127.0.0.1 logea DENIAL (otro rango privado)', async () => {
+    let err = null;
+    const stderr = await captureStderr(async () => {
+        try {
+            await get('http://127.0.0.1:1/', {
+                timeout: 5000,
+                agentTag: 'test-loopback',
+            });
+        } catch (e) {
+            err = e;
+        }
+    });
+    assert.equal(err && err.code, 'ERR_SSRF_BLOCKED');
+    assert.match(stderr, /DENIAL ERR_SSRF_BLOCKED/);
+    assert.match(stderr, /host=127\.0\.0\.1/);
+});
+
+test('CA-11 · request exitoso NO emite DENIAL', withPermissiveSSRF(async () => {
+    const srv = await localHttpServer((req, res) => {
+        res.writeHead(200); res.end('ok');
+    });
+    try {
+        const stderr = await captureStderr(async () => {
+            await get(srv.url + '/', { timeout: 5000, agentTag: 'test-happy' });
+        });
+        assert.ok(!/DENIAL/.test(stderr), `no debería haber DENIAL en stderr: ${stderr}`);
+    } finally {
+        await srv.close();
+    }
+}));

--- a/.pipeline/lib/__tests__/http-client.test.js
+++ b/.pipeline/lib/__tests__/http-client.test.js
@@ -523,3 +523,49 @@ test('CA-11 · request exitoso NO emite DENIAL', withPermissiveSSRF(async () => 
         await srv.close();
     }
 }));
+
+test('CA-11.1 · DENIAL sobre URL de Telegram NO filtra el BOT_TOKEN a stderr', async () => {
+    // Escenario real de takeover: el SSRF guard rechaza api.telegram.org porque
+    // el resolver DNS devuelve una IP privada (DNS rebinding / /etc/hosts
+    // manipulado / proxy mal configurado). El http-client.js emite DENIAL y
+    // loggea la URL. Si `redactUrlLike` no cierra el path `/bot<TOKEN>/...`,
+    // el token se escribe en claro a stderr → takeover total del bot.
+    //
+    // Este test fuerza ese path con un resolver DNS custom que mapea
+    // api.telegram.org → 127.0.0.1 (privada).
+    const FAKE_TOKEN = '1234567890:ABCDefGHIjklMNOpqrsTUVwxyz_sensitive';
+    const telegramUrl = `https://api.telegram.org/bot${FAKE_TOKEN}/sendMessage`;
+    const rebindingResolver = {
+        lookup: async (_host, _opts) => [{ address: '127.0.0.1', family: 4 }],
+    };
+
+    let err = null;
+    const stderr = await captureStderr(async () => {
+        try {
+            await get(telegramUrl, {
+                timeout: 5000,
+                agentTag: 'test-ca11-telegram',
+                _dnsResolver: rebindingResolver,
+            });
+        } catch (e) {
+            err = e;
+        }
+    });
+
+    // El error sigue propagándose (no fail-open).
+    assert.ok(err, 'el request debía fallar');
+    assert.equal(err.code, 'ERR_SSRF_BLOCKED');
+
+    // El DENIAL fue emitido...
+    assert.match(stderr, /DENIAL ERR_SSRF_BLOCKED/, `debió loggear DENIAL: ${stderr}`);
+    assert.match(stderr, /host=api\.telegram\.org/);
+
+    // ...pero el token NO debe aparecer en stderr, en ninguna variante.
+    assert.ok(!stderr.includes(FAKE_TOKEN),
+        `BOT_TOKEN filtrado a stderr: ${stderr}`);
+    assert.ok(!stderr.includes('ABCDefGHIjklMNOpqrsTUVwxyz_sensitive'),
+        `fragmento opaco del token filtrado: ${stderr}`);
+    // Debe aparecer el marker de redacción.
+    assert.match(stderr, /\/bot\[REDACTED\]/,
+        `path del bot debe estar redactado: ${stderr}`);
+});

--- a/.pipeline/lib/__tests__/redact.test.js
+++ b/.pipeline/lib/__tests__/redact.test.js
@@ -147,6 +147,42 @@ test('SEC-3 · URL con userinfo se strippea antes de loguear', () => {
     assert.ok(out.includes(`${REDACTION_MARKER}@`));
 });
 
+test('CA-11.1 · URL de Telegram con /bot<TOKEN> redacta el path (no fuga en DENIAL)', () => {
+    // Reproduce el caso real del DENIAL de http-client.js cuando el SSRF guard
+    // rechaza api.telegram.org: el path lleva el BOT_TOKEN completo. Si
+    // `redactUrlLike` solo redactaba userinfo+query, el token se escribía
+    // literal a stderr → takeover total del bot (#2332 rebote).
+    const url = 'https://api.telegram.org/bot1234567890:ABCDefGHIjklMNOpqrsTUVwxyz/sendMessage';
+    const out = redactUrlLike(url);
+    assert.ok(out.includes(REDACTION_MARKER), `esperaba marker en: ${out}`);
+    assert.ok(!out.includes('1234567890:ABCDefGHIjklMNOpqrsTUVwxyz'),
+        `token no debe estar en claro: ${out}`);
+    assert.ok(!out.includes('ABCDefGHIjklMNOpqrsTUVwxyz'),
+        `segmento del token no debe filtrarse: ${out}`);
+    // Debe preservar host + método para que el log siga siendo contextual.
+    assert.ok(out.includes('api.telegram.org'));
+    assert.ok(out.includes('/sendMessage'));
+});
+
+test('CA-11.1 · /bot<TOKEN> en distintas posiciones y query', () => {
+    // Path + query: el path debe redactarse Y los query params sensibles también.
+    const url1 = 'https://api.telegram.org/bot987654321:XYZabcDEF123ghiJKLmno/getUpdates?offset=5';
+    const out1 = redactUrlLike(url1);
+    assert.ok(!out1.includes('XYZabcDEF123ghiJKLmno'));
+    assert.ok(out1.includes('/bot[REDACTED]'));
+    assert.ok(out1.includes('offset=5'), 'offset no es sensible, debe sobrevivir');
+
+    // Token corto (<20) no matchea — evita falsos positivos tipo /bot/list.
+    const url2 = 'https://example.com/bot/list';
+    const out2 = redactUrlLike(url2);
+    assert.equal(out2, url2, 'path /bot/list sin token no debe tocarse');
+
+    // Case insensitive (Telegram acepta /Bot también).
+    const url3 = 'https://api.telegram.org/Bot1234567890:ABCDefGHIjklMNOpqrsTUVwxyz/sendMessage';
+    const out3 = redactUrlLike(url3);
+    assert.ok(!out3.includes('ABCDefGHIjklMNOpqrsTUVwxyz'));
+});
+
 test('redactSensitive · ciclos no revientan', () => {
     const a = { password: 'x' };
     a.self = a;

--- a/.pipeline/lib/http-client.js
+++ b/.pipeline/lib/http-client.js
@@ -203,11 +203,27 @@ const CROSS_ORIGIN_DROP_HEADERS = ['authorization', 'cookie', 'proxy-authorizati
  * @returns {Promise<{statusCode, headers, bodyBuffer, finalUrl}>}
  */
 async function doSingleRequest(parsedUrl, options, logger) {
+    // Precalculo de URL redactada para logs (CA-6/CA-11/CA-17).
+    const redactedUrl = redactUrlLike(parsedUrl.toString());
+
     // Validar SSRF antes de abrir socket (CA-9 + CA-13).
-    const ipList = await Promise.race([
-        ssrfGuard.validateHostname(parsedUrl.hostname, { dnsResolver: options._dnsResolver }),
-        rejectAfter(TIMEOUT_DNS_MS, ERROR_CODES.TIMEOUT_DNS, 'timeout de resolución DNS'),
-    ]);
+    // CA-11 (#2332): si el guard rechaza, logeamos DENIAL estructurado antes de
+    // propagar para que producción deje traza de intentos (URL, host, razón, stack).
+    let ipList;
+    try {
+        ipList = await Promise.race([
+            ssrfGuard.validateHostname(parsedUrl.hostname, { dnsResolver: options._dnsResolver }),
+            rejectAfter(TIMEOUT_DNS_MS, ERROR_CODES.TIMEOUT_DNS, 'timeout de resolución DNS'),
+        ]);
+    } catch (err) {
+        if (err && (err.code === ERROR_CODES.SSRF_BLOCKED || err.code === ERROR_CODES.PROXY_NOT_WHITELISTED)) {
+            const stackLine = err.stack ? err.stack.split('\n').slice(0, 5).join(' | ') : '(sin stack)';
+            logger.error(
+                `DENIAL ${err.code} url=${redactedUrl} host=${parsedUrl.hostname} razon=${err.message} stack=${stackLine}`
+            );
+        }
+        throw err;
+    }
 
     const pickedIp = ipList[0];
     const isHttps = parsedUrl.protocol === 'https:';
@@ -258,7 +274,6 @@ async function doSingleRequest(parsedUrl, options, logger) {
     }
 
     const agentTag = options._agentTag || 'pipeline';
-    const redactedUrl = redactUrlLike(parsedUrl.toString());
     logger.info(`${requestOptions.method} ${redactedUrl} → ${pickedIp.address}`);
 
     const requester = isHttps ? https.request : http.request;

--- a/.pipeline/lib/redact.js
+++ b/.pipeline/lib/redact.js
@@ -101,7 +101,9 @@ function redactEmailsInText(text) {
 /**
  * Enmascara query strings sensibles dentro de una URL como texto plano
  * (CA-17). Mantiene el resto de los params intactos. También strippea
- * userinfo (user:pass@host) antes de loguear (CA-18 / SEC-3).
+ * userinfo (user:pass@host) antes de loguear (CA-18 / SEC-3) y redacta
+ * secretos embebidos en el path (ej: `/bot<TOKEN>/sendMessage` del
+ * API de Telegram — CA-11.1 / #2332).
  *
  * No usa `URL` para evitar excepciones con strings inválidos (URLs en
  * error.message pueden estar truncadas).
@@ -125,6 +127,13 @@ function redactUrlLike(urlText) {
         }
         return m;
     });
+
+    // 3) Redactar secretos embebidos como path segments (CA-11.1 / #2332).
+    //    El API de Telegram usa /bot<TOKEN>/metodo — si este path se loggea
+    //    dentro de un DENIAL (SSRF guard rechazando por DNS rebind o proxy
+    //    mal configurado) el BOT_TOKEN se filtra a stderr. Redactamos todo
+    //    segmento que empieza con `/bot` seguido de 20+ chars opacos.
+    out = out.replace(/\/bot[^\/?#\s]{20,}/gi, `/bot${REDACTION_MARKER}`);
 
     return out;
 }

--- a/.pipeline/servicio-telegram.js
+++ b/.pipeline/servicio-telegram.js
@@ -2,11 +2,17 @@
 // =============================================================================
 // Servicio Telegram — Fire-and-forget message sender
 // Procesa cola de servicios/telegram/pendiente/
+//
+// Migrado a http-client seguro (issue #2332):
+//   - SSRF guard sobre api.telegram.org (CA-9/CA-13)
+//   - TLS estricto, timeouts escalonados, body cap, CRLF protect
+//   - Denials loggeados estructuradamente (CA-11)
 // =============================================================================
 
-const https = require('https');
 const fs = require('fs');
 const path = require('path');
+const httpClient = require('./lib/http-client');
+const { ERROR_CODES } = require('./lib/constants');
 
 const PIPELINE = process.env.PIPELINE_STATE_DIR || path.resolve(__dirname);
 const QUEUE_DIR = path.join(PIPELINE, 'servicios', 'telegram');
@@ -32,72 +38,93 @@ function log(msg) {
   console.log(`[${ts}] [svc-telegram] ${msg}`);
 }
 
-function telegramSend(method, params) {
-  return new Promise((resolve, reject) => {
-    const data = JSON.stringify({ chat_id: CHAT_ID, ...params });
-    const options = {
-      hostname: 'api.telegram.org',
-      path: `/bot${BOT_TOKEN}/${method}`,
-      method: 'POST',
-      headers: { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(data) }
-    };
-    const req = https.request(options, (res) => {
-      let body = '';
-      res.on('data', (c) => body += c);
-      res.on('end', () => resolve(JSON.parse(body)));
-    });
-    req.on('error', reject);
-    req.setTimeout(30000, () => { req.destroy(); reject(new Error('timeout')); });
-    req.write(data);
-    req.end();
-  });
+// Tag fijo para logs del http-client — permite filtrar denials del servicio.
+const AGENT_TAG = 'svc-telegram';
+
+/**
+ * Logging estructurado de denial SSRF/proxy (CA-11 del #2332).
+ * El http-client ya logea internamente, pero replicamos al log persistente
+ * del servicio para trazabilidad post-mortem (crash-handlers escriben acá).
+ */
+function logDenialIfAny(method, err) {
+  if (!err) return;
+  const code = err.code;
+  if (code === ERROR_CODES.SSRF_BLOCKED || code === ERROR_CODES.PROXY_NOT_WHITELISTED) {
+    log(`DENIAL ${code} method=${method} razon=${err.message}`);
+  }
 }
 
-/** Enviar documento/foto via multipart form-data */
-function telegramSendMultipart(method, fieldName, filePath, extra = {}) {
-  return new Promise((resolve, reject) => {
-    const boundary = '----PipelineV2' + Date.now();
-    const filename = path.basename(filePath);
-    const fileData = fs.readFileSync(filePath);
+/**
+ * Envío JSON a la API de Telegram vía http-client seguro.
+ * POST con `retryable:true` porque sendMessage de Telegram tolera duplicados
+ * con el mismo texto (Telegram de-dupea por chat_id + text cuando llega rápido).
+ */
+async function telegramSend(method, params) {
+  const url = `https://api.telegram.org/bot${BOT_TOKEN}/${method}`;
+  const body = { chat_id: CHAT_ID, ...params };
+  try {
+    const res = await httpClient.postJson(url, body, {
+      agentTag: AGENT_TAG,
+      timeout: 30000,
+      retryable: true, // idempotente en la práctica para sendMessage
+    });
+    return res.body;
+  } catch (err) {
+    logDenialIfAny(method, err);
+    throw err;
+  }
+}
 
-    let body = '';
-    // chat_id field
-    body += `--${boundary}\r\nContent-Disposition: form-data; name="chat_id"\r\n\r\n${CHAT_ID}\r\n`;
-    // extra fields (caption, parse_mode, etc.)
-    for (const [key, val] of Object.entries(extra)) {
-      body += `--${boundary}\r\nContent-Disposition: form-data; name="${key}"\r\n\r\n${val}\r\n`;
-    }
-    // file field
-    const fileHeader = `--${boundary}\r\nContent-Disposition: form-data; name="${fieldName}"; filename="${filename}"\r\nContent-Type: application/octet-stream\r\n\r\n`;
-    const fileFooter = `\r\n--${boundary}--\r\n`;
+/** Enviar documento/foto via multipart form-data usando http-client seguro. */
+async function telegramSendMultipart(method, fieldName, filePath, extra = {}) {
+  const boundary = '----PipelineV2' + Date.now();
+  const filename = path.basename(filePath);
+  const fileData = fs.readFileSync(filePath);
 
-    const bodyBuf = Buffer.concat([
-      Buffer.from(body + fileHeader),
-      fileData,
-      Buffer.from(fileFooter)
-    ]);
+  let prologue = '';
+  // chat_id field
+  prologue += `--${boundary}\r\nContent-Disposition: form-data; name="chat_id"\r\n\r\n${CHAT_ID}\r\n`;
+  // extra fields (caption, parse_mode, etc.)
+  for (const [key, val] of Object.entries(extra)) {
+    prologue += `--${boundary}\r\nContent-Disposition: form-data; name="${key}"\r\n\r\n${val}\r\n`;
+  }
+  // file field header
+  const fileHeader = `--${boundary}\r\nContent-Disposition: form-data; name="${fieldName}"; filename="${filename}"\r\nContent-Type: application/octet-stream\r\n\r\n`;
+  const fileFooter = `\r\n--${boundary}--\r\n`;
 
-    const options = {
-      hostname: 'api.telegram.org',
-      path: `/bot${BOT_TOKEN}/${method}`,
+  const bodyBuf = Buffer.concat([
+    Buffer.from(prologue + fileHeader),
+    fileData,
+    Buffer.from(fileFooter),
+  ]);
+
+  const url = `https://api.telegram.org/bot${BOT_TOKEN}/${method}`;
+  try {
+    const res = await httpClient.request(url, {
       method: 'POST',
+      body: bodyBuf,
       headers: {
         'Content-Type': `multipart/form-data; boundary=${boundary}`,
-        'Content-Length': bodyBuf.length
-      }
-    };
-    const req = https.request(options, (res) => {
-      let resp = '';
-      res.on('data', (c) => resp += c);
-      res.on('end', () => {
-        try { resolve(JSON.parse(resp)); } catch { resolve({ ok: false, description: resp }); }
-      });
+      },
+      agentTag: AGENT_TAG,
+      timeout: 60000,
+      // Envíos de archivo NO son idempotentes por lado de Telegram: NO retry automático.
+      // Cap de respuesta por default (10 MB) alcanza para la respuesta JSON del send.
     });
-    req.on('error', reject);
-    req.setTimeout(60000, () => { req.destroy(); reject(new Error('timeout')); });
-    req.write(bodyBuf);
-    req.end();
-  });
+    // La API devuelve JSON; si el parser del http-client no pudo parsearlo
+    // (p.ej. content-type no json), devolvemos estructura similar al código previo.
+    if (typeof res.body === 'string') {
+      try { return JSON.parse(res.body); } catch { return { ok: false, description: res.body }; }
+    }
+    if (Buffer.isBuffer(res.body)) {
+      const s = res.body.toString('utf8');
+      try { return JSON.parse(s); } catch { return { ok: false, description: s }; }
+    }
+    return res.body;
+  } catch (err) {
+    logDenialIfAny(method, err);
+    throw err;
+  }
 }
 
 function listWorkFiles(dir) {

